### PR TITLE
Adds server port environment option.

### PR DESCRIPTION
**Pull request by ncrmro**
Sometimes I find myself with more than one project open at once, it would be nice to be able to set a custom server port.

### CODE REVIEW

**Comment by mattkrick** [comment]
very reasonable, thanks!
could you add this to `getGraphQLHost` as well so it tries the env port before it defaults to 3000?

**Comment by patrickleet** [comment]
to access environment variables from the client you can use 

```
plugins: [
    //...
    new webpack.EnvironmentPlugin([
      'ENV_VAR_NAME'
    ])
  ],
```

Then you can just use ```process.env.ENV_VAR_NAME```

**Comment by ncrmro** [comment]
@patrickleet This works except it stops the build if the env variables aren't defined. Please let me know if there is a more elegant solution.

**Comment by patrickleet** [comment]
The builds work for me - it does throw a warning and look ugly though

**Comment by patrickleet** [comment]
@ncrmro also seems like most supported and current seems how's webpack's own solution. https://github.com/webpack/docs/wiki/list-of-plugins - too bad I don't see anything about default values

File: webpack/webpack.config.prod.js [hunk 1]
```diff
@@ -68,9 +72,12 @@ export default {
     new AssetsPlugin({path: path.join(root, 'build'), filename: 'assets.json'}),
     new webpack.NoErrorsPlugin(),
     new webpack.DefinePlugin({
-      '__CLIENT__': true,
+      '__CLIENT__': false,
```
Comment (mattkrick) [inline webpack/webpack.config.prod.js]: doesn't this break everything wrt SSR?

**Comment by mattkrick** [comment]
Neat! That `EnvironmentPlugin` looks like the best tool for the job if we wanna have a single variable for the port on client & server. If Meatier defines all those variables on the server then it should work by just replacing the static stuff with the `process.env.X`, right? Granted it's not super pretty, but having 1 variable source shared across client & server does seem awesome

File: webpack/webpack.config.prod.js (hunk 1 above)
Comment (ncrmro) [inline webpack/webpack.config.prod.js]: Yep, definitely an accident.

**Comment by ncrmro** [comment]
Any ideas better than this and dropping it into the server file, then using the EnviormentPlugin? 

```javascript
if (process.env.SERVER_PORT) {} else {
  process.env['SERVER_PORT'] = '3000';
}

if (process.env.GRAPHQL_HOST) {} else {
  process.env['GRAPHQL_HOST'] = 'localhost';
}

if (process.env.GRAPHQL_PROTOCOL) {} else {
  process.env['GRAPHQL_PROTOCOL'] = 'http:';
}
```

**Comment by patrickleet** [comment]
I'm using it for inserting a basename to be used with nginx url rewriting, and the EnvironmentPlugin has been incredibly useful for that. It was useful because I needed to access the variable in the webpack configs, the server, and the client.

**Comment by ncrmro** [comment]
Here is some proposed changes, still kinda clunky. I can create a pull request without the bad commits.

File: webpack/webpack.config.prod.js [hunk 2]
```diff
@@ -74,11 +70,13 @@ export default {
...
-      'process.env.SERVER_PORT': JSON.stringify(serverPort) || JSON.stringify('3000'),
-      'process.env.GRAPHQL_HOST': JSON.stringify(graphQLHost) || JSON.stringify('localhost'),
-      'process.env.GRAPHQL_PROTOCOL': JSON.stringify(graphQLProtocol) || JSON.stringify('http:')
-    })
+      'process.env.NODE_ENV': JSON.stringify('production')
+    }),
+    new webpack.EnvironmentPlugin([
+      'SERVER_PORT',
+      'GRAPHQL_HOST',
+      'GRAPHQL_PROTOCOL'
```
Comment (patrickleet) [inline webpack/webpack.config.prod.js]: you can pass NODE_ENV in here too - wondering if that would make one in webpack.DefinePlugin obsolete?

File: src/server/server.js [hunk 3]
```diff
@@ -5,13 +5,25 @@ import path from 'path';
...
+}
+
 export const options = {
   authKey: jwtSecret,
   logLevel: 1,
   // change this to scale vertically
   workers: 1 || numCpus,
   brokers: 1,
-  port: process.env.SERVER_PORT || 3000,
+  port: process.env.SERVER_PORT,
```
Comment (patrickleet) [inline src/server/server.js]: what's wrong with 
```|| 3000``` ?

Comment (ncrmro) [inline src/server/server.js]: The SERVER_PORT env var needs to be defined anyway, if it isn't the build process stops cause they are undefined for later use in the graphQL fetching.

**Comment by mattkrick** [comment]
how about eg `process.env.SERVER_PORT = process.env.SERVER_PORT || '3000';`?
ideally though, is there a `.env` that we could incorporate that is deploy agnostic? If so, we could just put that in the core meatier stack so the var is guaranteed to exist.

**Comment by ncrmro** [comment]
@mattkrick that definitely works, totally overthrought that one.

What are your thoughts on [dotenv](https://www.npmjs.com/package/dotenv) which will automatically pull in the .env and have a .sample-env intially checked in?

[Outlined here Managing Enviorment in Node JS](https://medium.com/@rafaelvidaurre/managing-environment-variables-in-node-js-2cb45a55195f#.sj08ntaq5)

**Comment by mattkrick** [comment]
yeah, i'm all for it. honestly though, i'm not a devops guy so i dont know if there's some deployment strategy that doesn't use that. @Bartekus any thoughts on including a .env? Would it negatively impact any deploy strategies that you've come across?

**Comment by mattkrick** [comment]
for Action (a project built on meatier) we went ahead & dropped in a .env file & used dotenv and dontenv-expand: https://github.com/ParabolInc/action/blob/9aa8f0dea8ba99ca655440b05c52e3a0c68a2547/src/server/server.js#L18

i think it's pretty great looking & should probably be adopted here, too. Thoughts?

**Comment by ncrmro** [comment]
I've attempted to integrate these changes here #136 .

**Comment by jedwards1211** [comment]
since the app could be running behind a proxy with a different port I think `PORT` should be used by `server.js` only, and a separate `ROOT_URL` variable be used on the client (still injected with `EnvironmentPlugin`).

**Pull request closed without merging**